### DISCUSSION TITLE
Help with creating an Arduino Opta target

### DISCUSSION THREAD

**Discussion by putertubby** [mbed-ce/mbed-os#313] [Q&A] [answered]
# Background
We're currently working on software for the [Arduino Opta](https://docs.arduino.cc/hardware/opta/). The Arduino core is based on Mbed OS but Arduino libraries don't support asynchronous I/O. We tried a multi-threaded approach but it turned out very bulky so we opted to ditch the Arduino libraries altogether and write native code for the Mbed OS-based Arduino core supporting asynchronous I/O instead. This approach works very well but the Arduino build system comes with some severe limitations so we were looking to get rid of the Arduino core and `arduino-cli` as well and build the application on pure Mbed OS using the the Mbed OS build system instead. However, due to [recent announcements](https://forums.mbed.com/t/important-update-on-mbed-end-of-life/23644) we are reconsidering that idea and are looking into porting over to Mbed CE as an alternative.
# Question
Neither Mbed OS nor Mbed CE support Arduino Opta targets out-of-the box. However, I'm told that the [Arduino Portenta H7](https://forums.mbed.com/t/important-update-on-mbed-end-of-life/23644), which is supported by Mbed CE, is very similar to the Opta. That assumption seems to be confirmed by the very few differences I find when comparing the Portenta and Opta targets in the Mbed OS part f the Arduino core.

~~~
$ diff --recursive TARGET_PORTENTA_H7/ TARGET_OPTA/
diff --recursive TARGET_PORTENTA_H7/COMPONENT_WHD/whd_config.h TARGET_OPTA/COMPONENT_WHD/whd_config.h
51c51
< //#define WIFI_32K_CLK      {GPIOA,{.Pin= GPIO_PIN_8, .Mode = GPIO_MODE_AF_PP , .Pull = GPIO_NOPULL , .Speed= GPIO_SPEED_FREQ_LOW , .Alternate = GPIO_AF0_MCO}}
---
> //#define WIFI_32K_CLK        {GPIOA,{.Pin= GPIO_PIN_6, .Mode = GPIO_MODE_AF_PP , .Pull = GPIO_NOPULL , .Speed= GPIO_SPEED_FREQ_LOW , .Alternate = GPIO_AF0_MCO}}
Only in TARGET_OPTA/: opta_power.h
diff --recursive TARGET_PORTENTA_H7/PinNames.h TARGET_OPTA/PinNames.h
333c333
<     CONSOLE_TX = PA_9,
---
>     CONSOLE_TX = PG_14,
338c338
<     CONSOLE_RX = PA_10,
---
>     CONSOLE_RX = PG_9,
342,349c342,354
<     LED_RED = PK_5,  //Red
<     LED_GREEN = PK_6,  //Green
<     LED_BLUE = PK_7,  //Blue
< 
<     CYBSP_BT_UART_RX = PF_6,
<     CYBSP_BT_UART_TX = PA_15,
<     CYBSP_BT_UART_RTS = PF_8,
<     CYBSP_BT_UART_CTS = PF_9,
---
>     LED_RED = PH_11,  //Red
>     LED_GREEN = PH_12,  //Green
>     //LED_BLUE = PK_7,  //Blue
> 
>     LED_SYS_1 = PH_11,
>     LED_SYS_2 = PH_12,
>     LED_BLE = PE_5,
> 
> 
>     CYBSP_BT_UART_RX = PH_14,
>     CYBSP_BT_UART_TX = PB_9,
>     CYBSP_BT_UART_RTS = PA_15,
>     CYBSP_BT_UART_CTS = PB_15,
358c363
<     QSPI_FLASH1_IO2 = PF_7,
---
>     QSPI_FLASH1_IO2 = PE_2,
360c365
<     QSPI_FLASH1_SCK = PF_10,
---
>     QSPI_FLASH1_SCK = PB_2,
410c415
<     SYS_JTDO_SWO = PB_3,
---
>     SYS_JTDO_SWO = PC_6,
Only in TARGET_PORTENTA_H7/: portenta_power.h
$ diff TARGET_PORTENTA_H7/portenta_power.h TARGET_OPTA/opta_power.h
17,18c17,18
< #ifndef PORTENTA_POWER
< #define PORTENTA_POWER
---
> #ifndef OPTA_POWER
> #define OPTA_POWER
$
~~~
I am a complete noob at Mbed OS and know next to nothing about Mbed CE so I can't say if this is all that differs but can't find any other obvious places to look.

Finally, here are my questions:
1. Are the above differences likely all there is or is there probably (or certainly) more to be found? Where should I look for more differences?
3. Assuming the above is everything; is it realistic to create an Arduino Opta target for Mbed CE based on the Arduino Portenta H7 target?
4. How would one go about creating a target for Arduino Opta based on the Arduino Portenta H7?
# Post Scriptum
I want to end with conveying my appreciation to the people working on this project, cloning and building `mbed-ce-hello-world` was such a nice experience! I really appreciate the CMake build system.

**Comment by JohnK1987** [comment]
Hello and welcome to MbedCE community,

how you wrote the Opta is based on STM32H747xI so from first view i can say you should be able to start with example of [Custom target](https://github.com/mbed-ce/mbed-ce-custom-targets) project and it seems all necessary settings are available in Arduino Mbed Core repo.

So when you transform sources from [here](https://github.com/arduino/ArduinoCore-mbed/tree/main/variants/OPTA) to the Custom target project you are the winner i think.

Do not be hesitate to ask.

BR, Jan

**Comment by putertubby** [comment]
Thanks for answering!
I did some pattern-matching and came up with a game-plan, any and all feedback is appreciated!
1. Create a directory called "custom_target/ARDUINO_OPTA"
2. In that directory I create a file `CMakeLists.txt` with the following content (in principle a copy of the Portenta H7 file):
~~~
add_library(mbed-arduino-opta-m4 INTERFACE)

target_sources(mbed-arduino-opta-m4
    INTERFACE
        opta_power.cpp
)

target_include_directories(mbed-arduino-opta-m4
    INTERFACE
        .
)

target_link_libraries(mbed-arduino-opta-m4 INTERFACE mbed-stm32h747xi-cm4)


add_library(mbed-arduino-opta-m7 INTERFACE)

target_sources(mbed-arduino-opta-m7
    INTERFACE
        opta_power.cpp
)

target_include_directories(mbed-arduino-opta-m7
    INTERFACE
        .
)

target_link_libraries(mbed-arduino-opta-m7 INTERFACE mbed-stm32h747-arduino)
~~~
3. Copy the file `PinNames.h` from [here](https://github.com/arduino/ArduinoCore-mbed/blob/main/cores/arduino/mbed/targets/TARGET_STM/TARGET_STM32H7/TARGET_STM32H747xI/TARGET_OPTA/PinNames.h) and the file `opta_power.h` from the same location to the created directory.
4. Edit the file `CMakeLists.txt` in the `custom_targets` directory, add the line `add_subdirectory(ARDUINO_OPTA)`
5. Create the file `custom_targets.json5` with the following content (again just a copy of the Portenta targets):
~~~
{
    "ARDUINO_OPTA": {
        "public": false,
        "detect_code": ["0813"],
        inherits: ["STM32H747_ARDUINO"],
        "device_has_add": [
            "EMAC"
        ],
        "overrides": {
            "system_power_supply": "PWR_SMPS_1V8_SUPPLIES_LDO",
            "clock_source": "USE_PLL_HSE_EXTC",
            "lse_bypass": 1,
            "network-default-interface-type": "ETHERNET",
            "hse_value": 25000000,
            // Can use overdrive because LDO is regulating core voltage (even though
            // the LDO power comes from the SMPS).
            "enable-overdrive-mode": 1
        },
        "device_name": "STM32H747XIHx",
        "image_url": "https://store.arduino.cc/cdn/shop/products/ABX00042_00.iso_1200x900.jpg?v=1675840144"
    },
    "ARDUINO_OPTA_M7": {
        "inherits": ["ARDUINO_OPTA"],
        "core": "Cortex-M7FD",
        "mbed_rom_start": "0x08000000",
        "mbed_rom_size" : "0x100000",
        "mbed_ram_start": "0x24000000",
        "mbed_ram_size" : "0x80000",
        "extra_labels_add": [
            "STM32H747xI_CM7"
        ],
        "macros_add": [
            "CORE_CM7"
        ]
    },
    "ARDUINO_OPTA_M4": {
        "inherits": ["ARDUINO_OPTA"],
        "core": "Cortex-M4F",
        "mbed_rom_start": "0x08100000",
        "mbed_rom_size" : "0x100000",
        "mbed_ram_start": "0x10000000",
        "mbed_ram_size" : "0x48000",
        "extra_labels_add": [
            "STM32H747xI_CM4"
        ],
        "macros_add": [
            "CORE_CM4",
            "MBED_MPU_CUSTOM"
        ],
        "device_has_remove": [
            "MPU"
        ]
    }
}
~~~
6. Update the memory map in the file above to match the expectations of the Opta boot (can I keep the original boot or do I need to change the boot to use Mbed CE by the way?). It seems that the M7 core flash starts at 0x8040000 according to the [linker script](https://github.com/arduino/ArduinoCore-mbed/blob/main/variants/OPTA/linker_script.ld).

The only missing piece(?) now is the implementation of opta_power (`opta_power.cpp`), any suggestions on what to do with that? Will the portenta_power file work if renamed?

Making the above changes allows me to configure and build the Custom Target project like this:
~~~
$ cmake -B build -S . -GNinja -DCMAKE_BUILD_TYPE=Develop -DMBED_TARGET=ARDUINO_OPTA_M7
$ cmake --build build --target all --parallel
~~~

**Reply by JohnK1987** [reply, marked as answer]
First, you can use **Write a reply** under every answer and not **Suggest an answer**. That will allow you to lead conversations with multiple contributors about one topic separately.

----------------------------------------------------------------------------------------------

About point 5. - it depends on your target and your requirements, then you should to change this according to your requirements. 
* https://github.com/arduino/ArduinoCore-mbed/blob/1399d6477548f3f5585155652511d68d3ef85951/variants/OPTA/conf/mbed_app.json#L13
* https://github.com/arduino/ArduinoCore-mbed/blob/main/variants/OPTA/cxxflags.txt


----------------------------------------------------------------------------------------------

> The only missing piece(?) now is the implementation of opta_power (opta_power.cpp), any suggestions on what to do with that? Will the portenta_power file work if renamed?

If Opta's ethernet module is the same as Portenta's, then probably yes.

BR, Jan

**Reply by putertubby** [reply]
Thank you for the pointer, that makes sense!

I'm starting to connect the dots now, all the answers are in ArduinoCore-mbed. I just didn't look in the patch files at first. Also thanks for pointing me to the mbed_app.json file. I started reading up on the Mbed OS configuration process and now I believe that I have a rough idea about what's going on.

Here is how I understand the process on a very high level: The target properties are completed or overridden by the app settings in mbed_app.json and various files are generated. What threw me off initially was that these generated files are mixed in with the configuration files in the Arduino Core. One key such generated file is `mbed_config.h` in Mbed OS which seems to be replaced by `mbed-target-config.h` in Mbed CE? At least it seems that way based on the snippets below?

ArduinoCore-mbed:mbed.h
~~~
#ifndef MBED_H
#define MBED_H

#include "platform/mbed_version.h"
#include "mbed_config.h"

...
~~~

mbed-ce-custom-targets:mbed.h
~~~
#ifndef MBED_H
#define MBED_H

// In some cases, Mbed headers are included from source files that don't have
// mbed-target-config.h applied via "-include" flag (this is true for the Arduino Mbed core).
// In that case, we should manually include the target config header here.
#include "mbed-target-config.h"

#include "platform/mbed_version.h"

...
~~~

For people reading this conversation in the future: The ArduinoCode-mbed patch files contains the code for `opta_power.cpp` and from what I can tell the function `enableEthPowerSupply()` is empty. I.e. it is different from the Arduino Portenta H7 version.

Anyway, the game plan right now is to compare the generated files in mbed-ce-custom-targets with the corresponding files in ArduinoCore-mbed and change the target- and/or app settings until I'm happy. I'll allso have a look through the patch files to make sure I get everything I need.

Thanks for the help, the pointers saved me time and effort!

**Reply by multiplemonomials** [reply]
Yes, that's about correct!  We changed the header generation process in Mbed CE a bit, so that pretty much all the #defines are in the header instead of passed on the compiler command line.  Ultimately, all of those settings come from the targets.json, mbed_app.json, and mbed_lib.json files.

In Mbed CE, we have also refactored the target config quite a bit vs the original Arduino patches.  I made a "super-target" called `STM32H747_ARDUINO` that contains all the common stuff for opta/portenta/giga.

**Reply by multiplemonomials** [reply]
Also, I looked at the Arduino patch file, and I took a crack at converting the target definition:
```
    "ARDUINO_OPTA": {
        "inherits": ["STM32H747_ARDUINO"],
        "config": {
            "lse_bypass": {
                "help": "1 to use an oscillator (not a crystal) on 32k LSE",
                "value": "1"
            }
        },
        "overrides": {
            "hse_value": 25000000,
            "usb_speed": "USE_USB_OTG_FS",
            "system_power_supply": "PWR_SMPS_1V8_SUPPLIES_LDO",
            "clock_source": "USE_PLL_HSE_EXTC",
            "lse_available": 1,
            "network-default-interface-type": "ETHERNET",
        },
        "macros_add": [
            "MBEDTLS_FS_IO",
            "MBEDTLS_SHA1_C"
        ],
        "device_has_add": [
            "EMAC",
        ],
        "image_url": "https://store.arduino.cc/cdn/shop/products/AFX00003_01.iso_1200x900.jpg?v=1696337866"
    },
```
You do have to be a bit careful because it does use some slightly different options to the Portenta

**Reply by putertubby** [reply]
Thanks, that most likely saved me a few grey hairs!

**Reply by putertubby** [reply]
I have had some success porting the application to a Mbed CE custom target based on the answers above but now I am a bit stuck again. Here are the questions I'm wrestling with currently.

1. I can compile the application but fail to link with the following error message.
~~~
[build] /usr/local/gcc-arm/bin/../lib/gcc/arm-none-eabi/13.3.1/../../../../arm-none-eabi/bin/ld: build/_deps/mbed-os-src/connectivity/drivers/emac/libmbed-emac.a(stm32xx_emac.cpp.obj): in function `STM32_EMAC::low_level_init_successful()':
[build] /home/a/b/build/_deps/mbed-os-src/connectivity/drivers/emac/TARGET_STM/stm32xx_emac.cpp:366:(.text._ZN10STM32_EMAC25low_level_init_successfulEv+0x4): undefined reference to `stm32_eth_init_weak_symbol_helper'
[build] collect2: error: ld returned 1 exit status
~~~
There is quite a bit of ethernet initialization code in the Arduino core patch file for Opta. My questions is: What is the canonical way to solve the missing symbol? Should I apply the same changes as in the patch file or is there a way to provide the missing symbol and utilize the existing emac driver?

2. I'm using `dfu-util`  to program the Opta over USB since there is no debug port available. However, I can't find an upload method for `dfu-util` - is that correct? Is there a preferred way to provide a "custom" upload method for dfu-util?

3. Not specific for Mbed CE but rather a general Mbed question: I need to adjust the pullup/pulldown mode of one pin and the solution I found was including `pinmap.h` and using `pin_mode()`. I'm a bit confused by the multitude of API:s and abstraction layers. Is `pin-mode()` part of Mbed abstraction or is it STM-specific? I there a better (more general) way to do the same thing?

4. What is the best way to set the define `MBED_NO_GLOBAL_USING_DIRECTIVE`? Can I do that in `mbed_app.json5`?

**Reply by multiplemonomials** [reply]
1. Ah yes, this is due to a change I had to make to the Ethernet code to get it to link properly under CMake.  Head over to `stm32_eth_init.c` and drop the following code in it at the end:
```cpp
// Blank, non-weak-override function to make sure the linker pulls in this file
void stm32_eth_init_weak_symbol_helper()
{}
```

2. Yeah, currently we don't have dfu-util support, but I am planning to add it within the next few weeks-months.
3. The most friendly way to do that is to create an `mbed::DigitalIn` object and then either pass a PinMode to the constructor (e.g. `DigitalIn myPin(PX_y, PullUp)` or call the mode() function passing a pin mode.
4. Yeah, you should be able define an mbed_app.json like:
```json
{
    "macros_add": [
        "MBED_NO_GLOBAL_USING_DIRECTIVE"
    ]
}
```

**Reply by putertubby** [reply]
1. Thanks, that did it! Now the application is running successfully on Mbed CE on the M7 core in an Arduino Opta, including TCP/TLS over ethernet, USB serial and RS485 modbus communications! :)

2. Only wanted to make sure I didn't miss anything. I'll try to come up with a dfu-utils upload method, if I succeed I'll post here. You might beat me to it though.

3. I tried that but had no luck. The pin I want to set to mode for is a serial Rx pin, maybe that changes things? The `pin_mode()` function works fine for now, when/if I encounter issues I'll dig deeper.

4. I tried that but could only make it work in the `custom_targets.json5` file until I realized (encouraged by your answer) that I had to put the `macros_add` section in the right place. Now it works! Here is an example `mbed_app.json5` in case anyone else needs it.

~~~
{
    "target_overrides": {
        "*": {
            "platform.stdio-baud-rate": 115200,
            "platform.stdio-buffered-serial": 1,
            "macros_add": [
                "MBED_NO_GLOBAL_USING_DIRECTIVE"
            ]
        }
    }
}
~~~

Thanks again for all support!

**Comment by multiplemonomials** [comment]
By the way, I have been working on this branch for some time: https://github.com/mbed-ce/mbed-os/pull/316/files
It might help point you in the right direction of stuff that needs to be changed, though there is still a lot more for me to do on it!

**Reply by putertubby** [reply]
Thanks for the answer, that is impressive work! I'll be sure to check it out even though I hope to get away with far less work than that. ;)

**Reply by multiplemonomials** [reply]
yeah, this one is a little more complex because I had to do a lot to get the wi-fi support working
